client.py

```diff
@@ -260,26 +294,70 @@ def close(self):
     # are joined into a line. Blank lines are keepalive packets which reset
     # the timeout: _readline() pauses until a complete line has been received.
     async def _readline(self):
-        line = b''
+        line = None
+        preheader = None
+        header = None
         start = utime.ticks_ms()
         while True:
-            if line.endswith(b'\n'):
-                self._ok = True  # Got at least 1 packet
-                if len(line) > 1:
-                    return line
-                line = b''
+            if preheader is None:
+                cnt = 10
+            elif header is None and preheader[1] != 0:
+                cnt = preheader[1] * 2
+            elif line is None:
+                cnt = (preheader[3] << 8) | preheader[2]
+                if cnt == 0:
+                    line = b""
+                    cnt = 1  # new-line termination missing
+            else:
+                cnt = 1  # only newline-termination missing
+            d = await self._read_small(cnt, start)
+            # d is not None and print("read small got", d, cnt)
+            if d is None:
+                if line is not None:
+                    return preheader, header, line.decode()
+                line = None
+                preheader = None
+                header = None
                 start = utime.ticks_ms()  # Blank line is keepalive
                 if self._led is not None:
                     self._led(not self._led())
-            d = self._sock.readline()
+                continue
+            if preheader is None:
+                preheader = bytearray(ubinascii.unhexlify(d))
+            elif header is None and preheader[1] != 0:
+                header = bytearray(ubinascii.unhexlify(d))
+            elif line is None:
+                line = d
+            else:
+                raise OSError  # got unexpected characters instead of \n
+
+    async def _read_small(self, cnt, start):
```

Comment:
Not reading a complete line but the parts preheader, header and message in seperate pieces as this keeps the resource usage to a minimum as not separation is required after receiving the line. (Your approach has to split the mid header and the application typically has to split the new-line termination resulting in 2 reallocations).

New-line termination still marks the end of a message.